packages/itwinui-react/src/core/utils/components/Popover.tsx

```diff
@@ -4,162 +4,273 @@
+  useDismiss,
+  useInteractions,
+  size,
+  autoUpdate,
+  offset,
+  flip,
+  shift,
+  autoPlacement,
+  inline,
+  hide,
+  FloatingFocusManager,
+  useHover,
+  useFocus,
+  safePolygon,
+  useRole,
+} from '@floating-ui/react';
+import type { SizeOptions, Placement } from '@floating-ui/react';
+import {
+  Box,
+  cloneElementWithRef,
+  useControlledState,
+  useMergedRefs,
+} from '../index.js';
+import type { PolymorphicForwardRefComponent } from '../index.js';
+import { Portal } from './Portal.js';
+import type { PortalProps } from './Portal.js';
+import { Surface } from '../../Surface/index.js';
+import { ThemeProvider } from '../../ThemeProvider/index.js';
+
+type PopoverOptions = {
+  /**
+   * Placement of the popover content.
+   * @default 'bottom-start'
+   */
+  placement?: Placement;
   /**
    * Controlled flag for whether the popover is visible.
    */
   visible?: boolean;
+
   /**
-   * Determines the events that cause the popover to show.
-   * Should not be used when `visible` is set.
-   * @see [tippy.js trigger prop](https://atomiks.github.io/tippyjs/v6/all-props/#trigger)
+   * Callback invoked every time the popover visibility changes as a result
+   * of internal logic. Should be used alongside `visible` prop.
    */
-  trigger?: string;
+  onVisibleChange?: (visible: boolean) => void;
   /**
-   * Placement of the popover content.
-   * @default 'bottom-start'
-   * @see [tippy.js placement prop](https://atomiks.github.io/tippyjs/v6/all-props/#placement).
+   * If true, the popover will close when clicking outside it.
    */
-  placement?: Placement;
-} & Omit<TippyProps, 'placement' | 'trigger' | 'visible'>;
+  closeOnOutsideClick?: boolean;
+  /**
+   * Use an external element (stored in state) as the trigger.
+   */
+  reference?: HTMLElement | null;
+  /**
+   * Whether the popover should match the width of the trigger.
+   */
+  matchWidth?: boolean;
+};
 
-/**
- * Wrapper around [tippy.js](https://atomiks.github.io/tippyjs)
- * with pre-configured props and plugins (e.g. lazy mounting, focus, etc).
- * @private
- */
-export const Popover = React.forwardRef((props: PopoverProps, ref) => {
-  const [mounted, setMounted] = React.useState(false);
-  const themeInfo = React.useContext(ThemeContext);
-  const isDomAvailable = useIsClient();
-
-  const tippyRef = React.useRef<Element>(null);
-  const refs = useMergedRefs(tippyRef, ref);
-
-  // Plugin to allow lazy mounting. See https://github.com/atomiks/tippyjs-react#lazy-mounting
-  const lazyLoad = {
-    fn: () => ({
-      onShow: () => setMounted(true),
-      onHidden: () => setMounted(false),
-    }),
+// keep public api small to start with
+type PopoverInternalProps = {
+  /**
+   * autoUpdate options that recalculates position
+   * to ensure the floating element remains anchored
+   * to its reference element, such as when scrolling
+   * and resizing the screen
+   *
+   * https://floating-ui.com/docs/autoUpdate#options
+   */
+  autoUpdateOptions?: {
```

Comment:
Decided to keep autoUpdateOptions and middleware internal?

Reply:
yup